client/sysinfo/src/sysinfo.rs

```diff
@@ -32,14 +33,98 @@ use std::{
 	time::{Duration, Instant},
 };
 
+/// Throughput as measured in bytes per second.
+#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
+pub struct Throughput(f64);
+
+const KIBIBYTE: f64 = 1024.0;
+
+impl Throughput {
+	/// `f64` kibibyte/s to byte/s.
+	pub fn from_kibs(kibs: f64) -> Throughput {
+		Throughput(kibs * KIBIBYTE)
+	}
+
+	/// `f64` mebibyte/s to byte/s.
+	pub fn from_mibs(mibs: f64) -> Throughput {
+		Throughput(mibs * KIBIBYTE * KIBIBYTE)
+	}
+
+	/// `f64` gibibyte/s to byte/s.
+	pub fn from_gibs(gibs: f64) -> Throughput {
+		Throughput(gibs * KIBIBYTE * KIBIBYTE * KIBIBYTE)
+	}
+
+	pub fn as_f64(&self) -> f64 {
+		self.0
+	}
+
+	/// [`Self`] as number of kibibyte/s.
+	pub fn as_kibs(&self) -> f64 {
+		self.0 / KIBIBYTE
+	}
+
+	/// [`Self`] as number of mebibyte/s.
+	pub fn as_mibs(&self) -> f64 {
+		self.0 / (KIBIBYTE * KIBIBYTE)
+	}
+
+	/// [`Self`] as number of gibibyte/s.
+	pub fn as_gibs(&self) -> f64 {
+		self.0 / (KIBIBYTE * KIBIBYTE * KIBIBYTE)
+	}
+
+	/// Normalizes [`Self`] to use the larges unit possible.
+	fn normalize(&self) -> (f64, &'static str) {
+		let bs = self.0;
+
+		if bs >= KIBIBYTE * KIBIBYTE * KIBIBYTE {
+			(self.as_gibs(), "GiBs")
+		} else if bs >= KIBIBYTE * KIBIBYTE {
+			(self.as_mibs(), "MiBs")
+		} else {
+			(self.as_kibs(), "KiBs")
+		}
+	}
+}
+
+impl fmt::Display for Throughput {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		let (value, unit) = self.normalize();
+		write!(f, "{:.2?} {}", value, unit)
+	}
+}
+
+impl Serialize for Throughput {
+	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+	where
+		S: Serializer,
+	{
+		// NOTE I will replace 4 with the actual length.
+		let mut seq = serializer.serialize_map(Some(4))?;
+		let (value, unit) = self.normalize();
+		seq.serialize_entry(unit, &value)?;
+		seq.end()
+	}
+}
+
+impl<'de> Deserialize<'de> for Throughput {
+    fn deserialize<D>(deserializer: D) -> Result<Throughput, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+		Ok(Throughput(1 as f64))
+    }
+}
```

Comment:
1. The `Serialize` should be implemented as a standalone function and attached to the fields of `HwBench` with `#[serde(serialize_with = "path")]` (since it requires special handling of serialization where it's always in megabytes)
2. I don't think we actually need deserialization. (:

Reply:
This is old code, I updated it now. Are you sure we don't need deserialization? Without custom deserialization the code didn't compile, now that I added deserialization it should.

Reply:
Should the `HwBench` always be in megabytes?

Reply:
I think so. At least not generic deserialization. Some of @ggwpez 's code might need it, but it's probably more appropriate to handle it the same way as with the serialization here (that is, with an attribute). Can you point to the exact piece of code which doesn't compile when you remove deserialization?

Reply:
> Should the `HwBench` always be in megabytes?

Yes.

Reply:
https://github.com/paritytech/substrate/blob/952030cfa6f11be6aef938e5359064c4cf6b30a9/utils/frame/benchmarking-cli/src/machine/hardware.rs#L34

This didn't work without the custom deserialization.

Reply:
Ah yes, this is what @ggwpez was talking about. That should just use a custom deserializer attached with an attribute, just like the `HwBench`.